### PR TITLE
855 create bigquery adapter for dlt resources

### LINKED ISSUE #855: create BigQuery adapter for dlt resources
**Background**
Following the idea of adapter as used in Weaviate and Qdrant () we implement one for BigQuery. The idea of an adapter is to add destination specific table and column hinst using `x-<destination>-` table and column hints. In case of BigQuery we want to fully specify
- partitions
- clusters
- data retention
+ 2 more
Part of this ticket is to figure out the most popular settings of this kind and implement them. Ask our community for help :)

**Tasks**
1. * [x]  reuse Qdrant / Weaviate concept. Define and adapter that adds `x-` elements as columns definitions
2. * [x] design adapter function signature, write nice docstrings, place it in `dlt.destinations.adapters`
3. * [x] extend SQL builder in `bigquery.py` to interpret new hints.
4. * [x] add unit tests for to existing sql builder tests
5. * [x] add specialized adapter test that will create table with those hints
6. * [x] update the destination documentation

### PR DESCRIPTION
**Pull request by Pipboyguy**
### Description

- Closes #855

### Additional Context

This pull request introduces an initial version of a BigQuery adapter with `x-bigquery-table` and `x-bigquery-column` hints for specifying:

- Partitions  ✔️
- Clustering  ✔️
- Data retention policies  ✔️
- Rounding behavior  ✔️
- Table description ✔️

### CODE REVIEW

File: tests/load/bigquery/test_bigquery_table_builder.py [hunk 1]
```diff
@@ -371,3 +380,413 @@ def demo_source() -> DltResource:
...
+            assert hints_ddl
+
+        with c.execute_query(
+            "SELECT ddl FROM `INFORMATION_SCHEMA.TABLES` WHERE table_name = 'no_hints'"
+        ) as cur:
+            no_hints_ddl: str = cur.fetchone()[0]
+            ast_no_hints = sqlglot.parse_one(no_hints_ddl, read="bigquery")
+            assert no_hints_ddl
+
+    diff = sqlglot.diff(ast_no_hints, ast_hints)
```
Comment (Pipboyguy) [inline tests/load/bigquery/test_bigquery_table_builder.py:467]: It seems simpler and more reliable to use the bq client directly to inspect the generated schema like I did in other tests. Please advise

File: tests/load/bigquery/test_bigquery_table_builder.py [hunk 2]
```diff
@@ -371,3 +380,471 @@ def demo_source() -> DltResource:
...
+
+    assert some_data._hints["x-bigquery-table-expiration"] == pendulum.datetime(2030, 1, 1)  # type: ignore
+
+
+@pytest.mark.parametrize(
+    "destination_config",
+    destinations_configs(all_staging_configs=True, subset=["bigquery"]),
+    ids=lambda x: x.name,
+)
+def test_adapter_hints_table_expiration(destination_config: DestinationTestConfiguration) -> None:
```
Comment (Pipboyguy) [inline tests/load/bigquery/test_bigquery_table_builder.py:819]: Is this a valuable hint?

If I should implement it, where can I find a handle to the DDL.

File: tests/load/bigquery/test_bigquery_table_builder.py [hunk 3]
```diff
@@ -371,3 +380,471 @@ def demo_source() -> DltResource:
...
+
+    assert some_data._hints["x-bigquery-table-description"] == table_description  # type: ignore
+
+
+@pytest.mark.parametrize(
+    "destination_config",
+    destinations_configs(all_staging_configs=True, subset=["bigquery"]),
+    ids=lambda x: x.name,
+)
+def test_adapter_hints_table_description(destination_config: DestinationTestConfiguration) -> None:
```
Comment (Pipboyguy) [inline tests/load/bigquery/test_bigquery_table_builder.py:769]: Is this a valuable hint?

If I should implement it, where can I find a handle to the DDL.

File: dlt/common/validation.py [hunk 4]
```diff
@@ -74,21 +75,19 @@ def validate_dict(
     def verify_prop(pk: str, pv: Any, t: Any) -> None:
         # covers none in optional and union types
         if is_optional_type(t) and pv is None:
-            pass
-        elif is_union_type(t):
-            # pass if value actually is none
+            return
+        if is_union_type(t):
```
[the remark this answers is not in the document]
Comment (rudolfix) [reply dlt/common/validation.py:79]: @sultaniman this got fixed and now `is_union_type` works for both. right?

File: dlt/common/validation.py [hunk 5]
```diff
@@ -74,21 +75,19 @@ def validate_dict(
...
+            # pass if value is none
             union_types = extract_union_types(t, no_none=True)
             # this is the case for optional fields
             if len(union_types) == 1:
                 verify_prop(pk, pv, union_types[0])
             else:
                 has_passed = False
                 for ut in union_types:
-                    try:
+                    with contextlib.suppress(DictValidationException):
```
Comment (rudolfix) [inline dlt/common/validation.py:88]: nice!

File: dlt/common/validation.py [hunk 6]
```diff
@@ -148,14 +147,14 @@ def verify_prop(pk: str, pv: Any, t: Any) -> None:
...
-            # get dict key and value type from generic and process each k: v of the dict
+            # get a dict key and value type from generic and process each k: v of the dict.
             _, d_v_t = get_args(t)
             for d_k, d_v in pv.items():
                 if not isinstance(d_k, str):
                     raise DictValidationException(
                         f"In {path}: field {pk} key {d_k} must be a string", path, pk, d_k
                     )
-                verify_prop(pk + f"[{d_k}]", d_v, d_v_t)
+                verify_prop(f"{pk}[{d_k}]", d_v, d_v_t)
```
Comment (rudolfix) [inline dlt/common/validation.py:157]: thx for cleaning up this super old code

File: dlt/destinations/impl/bigquery/bigquery.py [hunk 7]
```diff
@@ -266,15 +275,23 @@ def _get_table_update_sql(
...
+            c["name"] for c in new_columns if c.get("cluster") or c.get(CLUSTER_HINT, False)
+        }:
+            sql[0] = sql[0] + "\nCLUSTER BY " + ", ".join(cluster_list)
         return sql
 
-    def _get_column_def_sql(self, c: TColumnSchema, table_format: TTableFormat = None) -> str:
-        name = self.capabilities.escape_identifier(c["name"])
-        return (
-            f"{name} {self.type_mapper.to_db_type(c, table_format)} {self._gen_not_null(c.get('nullable', True))}"
+    # noinspection PyTypedDict
```
Comment (rudolfix) [inline dlt/destinations/impl/bigquery/bigquery.py:284]: what's this? some kind of mypy hint?

File: dlt/extract/hints.py [hunk 8]
```diff
@@ -351,10 +339,52 @@ def _merge_keys(t_: TResourceHints) -> TPartialTableSchema:
 
         return partial
 
+    @staticmethod
+    def new_table_template(
```
Comment (rudolfix) [inline dlt/extract/hints.py:343]: looks like you had some serious git conflict? look up for `make_hints`. I removed this one and made it independent function. but it is exactly the same

TLDR;> remove this as use `make_hints`

File: pyproject.toml [hunk 9]
```diff
@@ -138,6 +138,7 @@ types-psutil = "^5.9.5.16"
 types-psycopg2 = "^2.9.21.14"
 cryptography = "^41.0.7"
 google-api-python-client = ">=1.7.11"
+sqlglot = {extras = ["rs"], version = "^20.11.0"}
```
Comment (rudolfix) [inline pyproject.toml:141]: kick this out if not used!

File: tests/load/bigquery/test_bigquery_table_builder.py (hunk 1 above)
Comment (rudolfix) [reply tests/load/bigquery/test_bigquery_table_builder.py:467]: you are free to kick that out! but let's say I'd ask you to add snowflake adapter. could you reuse part of this code? if not - you can kick it out

File: tests/load/bigquery/test_bigquery_table_builder.py (hunk 3 above)
Comment (rudolfix) [reply tests/load/bigquery/test_bigquery_table_builder.py:776]: I do not understand your question...

File: tests/load/bigquery/test_bigquery_table_builder.py (hunk 2 above)
Comment (rudolfix) [reply tests/load/bigquery/test_bigquery_table_builder.py:826]: hmmm? expiration is super valuable. we may even introduce a standardized hint (not bq specific) for data expiration

**Review by rudolfix — CHANGES_REQUESTED** [review]
this is really good and code quality is high! I really like the query builder test. I see two missing tests:

1. we are testing `apply_hints`. could you test `additional_table_hints`? does it work? if I call it twice does it merge additional hints or overwrite them (we typically merge dicts)
2. 1-2 simple unit tests for adapter itself could be cool. do we set correct hints? what if we call adapter twice?

File: tests/load/bigquery/test_bigquery_table_builder.py (hunk 1 above)
Comment (Pipboyguy) [reply tests/load/bigquery/test_bigquery_table_builder.py:467]: The **ast** will look different from snowflakes, since they represent partitions differently. I don't see reuse value here.

I am kicking it out as advised

Comment (Pipboyguy) [reply tests/load/bigquery/test_bigquery_table_builder.py]: Addressed in 1245d6d9d00d62b660af36e396fcb0524ad42c83

File: dlt/destinations/impl/bigquery/bigquery.py (hunk 7 above)
Comment (Pipboyguy) [reply dlt/destinations/impl/bigquery/bigquery.py:280]: Pycharm specific flags - apologies, removed!

File: pyproject.toml (hunk 9 above)
Comment (Pipboyguy) [reply pyproject.toml]: resolved in 1245d6d9d00d62b660af36e396fcb0524ad42c83

**Comment by Pipboyguy** [comment]
TODO: Documentation

File: dlt/extract/hints.py (hunk 8 above)
Comment (Pipboyguy) [reply dlt/extract/hints.py]: Removed in 7d04fe084b7fbf4a577a7ec6889f7efb61c3cb5b

Apologies! Need to brush up on my gitfu

**Comment by Pipboyguy** [comment]
@rudolfix Good catch! Thank you!

`additional_table_hints`  is covered by the following tests (which I have renamed for clarity):
- `test_additional_table_hints_parsing_table_description`
- `test_additional_table_hints_table_description`
- `test_additional_table_hints_parsing_table_expiration`
- `test_additional_table_hints_table_expiration`

Finishing up requested todos:

- [ ] Add unit test for multiple adapter calls
- [ ] Add unit test for `additional_table_hints` expected merge behavior
- [ ] PASS `test_additional_table_hints_table_description`
- [ ] PASS `test_additional_table_hints_table_expiration`
- [ ] Documentation

Kindly advised if I missed anything?

**Review by rudolfix — APPROVED** [review]
LGTM! thanks for fixing a lot of typos, docs and really good tests
